Added fisher-yates shuffle.

chronologos opened the pull request:
Profiled the shuffle vs naive implementation and it's much faster
(theoretically O(n) runtime).
[shuffle-exe.prof.txt](https://github.com/TikhonJelvis/inductive-mazes/files/1392442/shuffle-exe.prof.txt)

chronologos committed:
Added fisher-yates shuffle.

Profiled the shuffle vs naive implementation and it's much faster
(theoretically O(n) runtime).

chronologos committed:
Added graphs generated with random wall weights

chronologos committed:
kruskal mst with UnionFind

TikhonJelvis commented:
Looks great! Sorry I'm just getting around to this now.

One thing is that Kruskal's algorithm is a bit hard to follow. It would be great to have a quick overview of the algorithm as a comment—maybe just cribbed from Wikipedia :)—and more explicit variable names.

The shuffle code looks good. The only thing I'd worry about is using `//` repeatedly in a loop—I don't believe `//` is very efficient. Have you ever used [`modify`](https://hackage.haskell.org/package/vector-0.12.0.1/docs/Data-Vector.html#v:modify) from `Data.Vector`? It would let you do the whole thing in place while keeping a pure interface. I've found it really handy in the past. It would even make the code a bit neater, I believe.

chronologos committed:
Made changes according to comments

chronologos commented:
Makes sense, I changed the shuffle code and added comments to the Kruskal code. I actually only meant to push the shuffle code but forgot that any extra changes pushed to my fork would still come here. But might as well! Let me know if there are any extra changes I should make.

TikhonJelvis commented:
Yeah, looks good. Thanks for the contribution!
